runtime/src/main/cpp/SingleLockList.hpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_SINGLE_LOCK_LIST_H
+#define RUNTIME_SINGLE_LOCK_LIST_H
+
+#include <cstddef>
+#include <memory>
+#include <mutex>
+
+#include "CppSupport.hpp"
+#include "Utils.h"
+
+namespace kotlin {
+
+// TODO: Consider different locking mechanisms.
+template <typename Value, typename Mutex = SimpleMutex>
+class SingleLockList : private NoCopyOrMove {
+private:
+    struct Node;
+
+public:
+    class Iterator {
+    public:
+        explicit Iterator(Node* node) noexcept : node_(node) {}
+
+        Value& operator*() noexcept { return node_->value; }
+
+        Iterator& operator++() noexcept {
+            node_ = node_->next.get();
+            return *this;
+        }
+
+        bool operator==(const Iterator& rhs) const noexcept { return node_ == rhs.node_; }
+
+        bool operator!=(const Iterator& rhs) const noexcept { return node_ != rhs.node_; }
+
+    private:
+        Node* node_;
+    };
+
+    class Iterable : private NoCopy {
+    public:
+        explicit Iterable(SingleLockList* list) noexcept : list_(list), guard_(list->mutex_) {}
+
+        Iterator begin() noexcept { return Iterator(list_->root_.get()); }
+
+        Iterator end() noexcept { return Iterator(nullptr); }
+
+    private:
+        SingleLockList* list_;
+        std::unique_lock<Mutex> guard_;
+    };
+
+    template <typename... Args>
+    Value* Emplace(Args... args) noexcept {
+        auto node = kotlin::make_unique<Node>(args...);
+        auto* result = &node.get()->value;
+        LockGuard<Mutex> guard(mutex_);
+        if (root_) {
+            root_->previous = node.get();
+        }
+        node->next = std::move(root_);
+        root_ = std::move(node);
+        return result;
+    }
+
+    // You can only `Erase` `Value`s that were returned by `Emplace`. Trying
```

Comment:
Would this be more type safe if we returned pointer to `Node` from `Emplace` instead?

Reply:
Done, less magic is involved now.